### ISSUE TITLE
hope RapydScript file can excute like coffee-script

### ISSUE THREAD

**Issue by adousen** [atsepkov/RapydScript#95]
Though I know RapydScript is designed as a precompiler for browser-side javascirpt. Is there a plan to run rapydscript file directly, juse like coffee-script : 

```coffee-script
coffee /path/to/script.coffee  //excute the coffee-script file
```

And, I hope RapydScript can do like coffee-script ,require the ``*.coffee`` files in node.js.
```javascript
require("coffee-script/register")
require("./script.coffee")
```
If we can do like above, for example,  we will be able to use RapydScript to write config file(gulp.pyj) for gulp. This is amazing!

**Comment by kovidgoyal** [comment]
If you want to execute rs, you have two options

1) rapydscript -x -o /dev/null file.pyj (at least this is the case in my fork https://github.com/kovidgoyal/rapydscript-ng)
2) rapydscript file.py | node

Create an alias or shell script wrapper to make this easy.

As for using RS to write standalone JS modules that can be imported in other JS code -- it's not really something I care much about, but perhaps @atsepkov has some tips for that.

**Comment by atsepkov** [comment]
Both of the above work in the main fork too, a third alternative will be possible as well once #90 is closed:

    #!/usr/bin/env rapydscript -x

Add the following at the top of the source file and chmod it to a+x

**Comment by kovidgoyal** [comment]
@atsepkov Maybe -x should be changed to not output the compiled JS at all, and only run the compiled script silently. I had originally designed it to be a poor man's REPL, not a way to use RS as an interpreter.

**Comment by atsepkov** [comment]
Sure, I like that idea. We could even have it output only if -p flag is passed as well and be silent if it isn't.

**Comment by kovidgoyal** [comment]
I got rid of the --pretty flag, replaced it with the --ugly flag since IMO default output should not be minified. But since we now have a working REPL, I am fine with having -x always suppress the compiled output.

**Comment by kovidgoyal** [comment]
https://github.com/kovidgoyal/rapydscript-ng/commit/43b2022d3fb2aef6a0450a9892b466d92a22c367

**Comment by kovidgoyal** [comment]
If you want to see the compiled output as well, use the -o flag in addition to -x

**Comment by adousen** [comment]
Thanks , I have a Preliminary solutions, code just like the coffee-script/register. 

## we need add three file to rapydscript 

### 1. rapydscript/register.js
```javascript
require('./tools/register');
```

### 2. rapydscirpt/tools/compilefile.js
```javascirpt
var RapydScript = require("rapydscript");
var path = require("path");
var fs = require('fs');
var vm = require('vm');

var answer, TOPLEVEL;

function load(mod) {
    return require('../tools/' + mod); 
}

var filename = "testdo.pyj"

var base_path = path.normalize(path.join(path.dirname(module.filename), ".."));
var src_path = path.join(base_path, 'src');

var OUTPUT_OPTIONS = {
    beautify: false,
    private_scope: false,
    omit_baselib: false,
    comments: true,
};

function compile_single_file(err, code) {
    var output;
    if (err) {
        console.error("ERROR: can't read file: " + file);
        process.exit(1);
    }
	function parse_file(code, filename, toplevel, options) {
		//RapydScript options
		var rs_options = {
			filename: filename,
			toplevel: toplevel,
			readfile: fs.readFileSync,
			basedir: path.dirname(filename),
			libdir: path.join(src_path, 'lib'),
			auto_bind: options.auto_bind,
		};

	    return RapydScript.parse(code, rs_options);
	}

    TOPLEVEL = parse_file(code, filename, TOPLEVEL, OUTPUT_OPTIONS);
    try {
        output = RapydScript.OutputStream(OUTPUT_OPTIONS);
    } catch(ex) {
        if (ex instanceof RapydScript.DefaultsError) {
            console.error(ex.msg);
            console.error("Supported options:");
            console.error(ex.defs);
            process.exit(1);
        }
        throw ex;
    }

    OUTPUT_OPTIONS.baselib = RapydScript.parse_baselib(src_path, OUTPUT_OPTIONS.beautify);
    TOPLEVEL.print(output);
    output = output.get();

    answer = output
}


module.exports = function(filename, sourceMap) {
    var ex, raw;
    if (sourceMap == null) {
      sourceMap = false;
    }
    try{
        raw = fs.readFileSync(filename, 'utf8');
    }catch(e) {
       ex = e
    }
    compile_single_file(ex, raw)
    //console.log(answer);
    return answer
}
```

### 3. rapydscript/tools/register.js
```javascirpt
(function() {
  var RapydScript, Module, fs, binary, child_process, ext, findExtension, fork, i, len, loadFile, path, ref;

  RapydScript = require('rapydscript');
  compiler = require("./compilefile")
  child_process = require('child_process');
  Module = require('module');
  fs = require('fs');
  var vm = require('vm');

  var FILES = RapydScript.FILES 

  function load_global(file) {
    try {
        // var code = fs.readFileSync(file, "utf8");

        // We need to interpret scripts dynamically, and inject into the global environment
        // with those glabal vars and functions from individual scripts, is there any way? 
        // IF we can do it, we donnot need to set OUTPUT_OPTIONS.omit_baselib to be false.
        
    } catch(ex) {
        // XXX: in case of a syntax error, the message is kinda
        // useless. (no location information).
        console.error("ERROR in file: " + file + " / " + ex);
        process.exit(1);
    }
  }

  FILES.forEach(load_global);

  path = require('path');

  loadFile = function(module, filename) {
    var answer;
    answer = compiler(filename, false);
    return module._compile(answer, filename);
  };

  if (require.extensions) {
    ref = [".pyj",".rapyd"];
    for (i = 0, len = ref.length; i < len; i++) {
      ext = ref[i];
      require.extensions[ext] = loadFile;
    }
    findExtension = function(filename) {
      var curExtension, extensions;
      extensions = path.basename(filename).split('.');
      if (extensions[0] === '') {
        extensions.shift();
      }
      while (extensions.shift()) {
        curExtension = '.' + extensions.join('.');
        if (Module._extensions[curExtension]) {
          return curExtension;
        }
      }
      return '.js';
    };
    Module.prototype.load = function(filename) {
      var extension;
      this.filename = filename;
      this.paths = Module._nodeModulePaths(path.dirname(filename));
      extension = findExtension(filename);
      //console.log(extension)
      Module._extensions[extension](this, filename);
      return this.loaded = true;
    };
  }

  if (child_process) {
    fork = child_process.fork;
    binary = require.resolve('../bin/rapydscript');
    child_process.fork = function(path, args, options) {
      //if (helpers.isRapyd(path)) {
        if (!Array.isArray(args)) {
          options = args || {};
          args = [];
        }
        args = [path].concat(args);
        path = binary;
      //}
      return fork(path, args, options);
    };
  }

}).call(this);
```

## usage:  

> require ``.pyj`` modules, and execute it immediately

### one/path/test.pyj
``` javascirpt
RapydScript = require("rapydscript/register");
fs = require("./dosomething.pyj");
```

### one/path/do.pyj
``` javascirpt
path = require("path");

base_path = path.normalize(path.join(path.dirname(module.filename), ".."));

def doSomething():
	print("it's ok")
	print(base_path)

doSomething()
```

## Discussion
If we can interpret lib scripts dynamically, and inject into the global environment with those glabal vars and functions from individual scripts, I we donnot need to set OUTPUT_OPTIONS.omit_baselib to be false.
 is there any way?

**Comment by kovidgoyal** [comment]
You can get a list of exports in a rapydscript module from toplevel.exports. So all you need is to generate code at the bottom of every module, something like:

```js
toplevel.exports.forEach(function(name) {
   print('exports.' + name + ' = name');
});
```
I'm not sure how node's require hooking mechanism works, but you would have to somehow ebsure that all the compiled RS is run in the same wrapped context (that is with a single shared instance of the baselib)

**Comment by atsepkov** [comment]
#90 has been closed, as of 315924b0de23a732add8644347d77485ba0b7fba you can just include this shebang at the top and chmod the file to `a+x` to have it execute:

    /usr/bin/env rapydscript -x

The `-x` (`--execute`) option has also been modify to omit compiled JavaScript by default. To print AND execute the JavaScript add `-p` option.
